pkg/northbound/gnmi/v2/set.go

```diff
@@ -0,0 +1,326 @@
+		}
+		// TODO replace the following with fine grained RBAC using OpenPolicyAgent Regos
+		if err := utils.TemporaryEvaluate(md); err != nil {
+			log.Warn(err)
+			return nil, errors.Status(errors.NewUnauthorized(err.Error())).Err()
+		}
+	}
+	targetUpdates := make(mapTargetUpdates)
+	targetRemoves := make(mapTargetRemoves)
+
+	prefixTargetID := configapi.TargetID(req.GetPrefix().GetTarget())
+	targetInfo := targetInfo{}
+
+	extensions, err := extractExtensions(req)
+	if err != nil {
+		log.Warn(err)
+		return nil, errors.Status(errors.NewInvalid(err.Error())).Err()
+	}
+
+	if len(req.GetUpdate())+len(req.GetReplace())+len(req.GetDelete()) < 1 {
+		err = errors.NewInvalid("no updates, replace or deletes in SetRequest")
+		log.Warn(err)
+		return nil, errors.Status(err).Err()
+	}
+
+	// Update - extract targets and their models
+	for _, u := range req.GetUpdate() {
+		targetID := configapi.TargetID(u.Path.GetTarget())
+		if targetID == "" { //Try the prefix
+			targetInfo.targetID = prefixTargetID
+		} else {
+			targetInfo.targetID = targetID
+		}
+
+		modelPlugin, err := s.getModelPlugin(ctx, topoapi.ID(targetInfo.targetID))
+		if err != nil {
+			log.Warn(err)
+			return nil, errors.Status(err).Err()
+		}
+
+		targetUpdates[targetInfo.targetID], err = s.doUpdateOrReplace(ctx, req.GetPrefix(), u, targetUpdates, modelPlugin)
+		if err != nil {
+			log.Warn(err)
+			return nil, errors.Status(err).Err()
+		}
+	}
+
+	// Replace
+	for _, u := range req.GetReplace() {
+		targetID := configapi.TargetID(u.Path.GetTarget())
+		if targetID == "" { //Try the prefix
+			targetInfo.targetID = prefixTargetID
+		} else {
+			targetInfo.targetID = targetID
+		}
+
+		modelPlugin, err := s.getModelPlugin(ctx, topoapi.ID(targetInfo.targetID))
+		if err != nil {
+			log.Warn(err)
+			return nil, errors.Status(err).Err()
+		}
+
+		targetUpdates[targetInfo.targetID], err = s.doUpdateOrReplace(ctx, req.GetPrefix(), u, targetUpdates, modelPlugin)
+		if err != nil {
+			log.Warn(err)
+			return nil, errors.Status(err).Err()
+		}
+	}
+
+	//Delete
+	for _, u := range req.GetDelete() {
+		targetID := configapi.TargetID(u.GetTarget())
+		if targetID == "" { //Try the prefix
+			targetInfo.targetID = prefixTargetID
+		} else {
+			targetInfo.targetID = targetID
+		}
+		modelPlugin, err := s.getModelPlugin(ctx, topoapi.ID(targetInfo.targetID))
+		if err != nil {
+			log.Warn(err)
+			return nil, errors.Status(err).Err()
+		}
+
+		targetRemoves[targetInfo.targetID], err = s.doDelete(req.GetPrefix(), u, targetRemoves, modelPlugin)
+		if err != nil {
+			log.Warn(err)
+			return nil, errors.Status(err).Err()
+		}
+	}
+
+	transaction, err := newTransaction(targetInfo, extensions, targetUpdates, targetRemoves, userName)
+	if err != nil {
+		log.Warn(err)
+		return nil, errors.Status(err).Err()
+	}
+
+	err = s.transactions.Create(ctx, transaction)
+	if err != nil {
+		log.Warn(err)
+		return nil, errors.Status(err).Err()
+	}
+	ch := make(chan configapi.TransactionEvent)
```

Comment:
We should use `WithTransactionID` here to limit the `Watch` events only to those related to this particular transaction.